{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "Since use of NS-Forest for discovery of minimum marker gene combinations\n",
    "for cell type identification from single-cell RNA sequencing data is the\n",
    "first processing step of the NCBI Cell pipeline, the objective of this\n",
    "document is to:\n",
    "\n",
    "-   Run NS-Forest on an example human lung cell CELLxGENE dataset\n",
    "\n",
    "## Background\n",
    "\n",
    "This [NS-Forest\n",
    "Tutorial](https://nsforest.readthedocs.io/en/latest/tutorial.html) walks\n",
    "through all aspects of using NS-Forest.\n",
    "\n",
    "## Development environment\n",
    "\n",
    "See [Introduction.ipynb](Introduction.ipynb).\n",
    "\n",
    "### Jupyter Notebook\n",
    "\n",
    "Launch Jupyter Notebook from a terminal in which `.zshenv` has been\n",
    "sourced, and the virtual environment has been activated.\n",
    "\n",
    "### Emacs Org Mode\n",
    "\n",
    "Launch Emacs from a terminal in which `.zshenv` has been sourced, then\n",
    "evaluate this code block to activate the virtual environment:\n",
    "\n",
    "``` commonlisp\n",
    "(pyvenv-activate \"../../.venv\")\n",
    "```\n",
    "\n",
    "# Run NS-Forest on an example human lung cell CELLxGENE dataset\n",
    "\n",
    "Following the tutorial, we write a function that runs NS-Forest on an\n",
    "example human lung cell CELLxGENE dataset. We assume the dataset has\n",
    "been previously identified and downloaded (see\n",
    "[Chapter-01-CELLxGENE.org](Chapter-01-CELLxGENE.org)).\n",
    "\n",
    "To begin, we import modules, and assign module scope variables:"
   ],
   "id": "00e83e79-c980-41bd-987b-052d85d21455"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "results": "silent",
    "session": "shared",
    "tangle": "../py/NSForest.py"
   },
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import nsforest as ns\n",
    "from nsforest import nsforesting\n",
    "import scanpy as sc\n",
    "\n",
    "DATA_DIR = \"../data\"\n",
    "\n",
    "CELLXGENE_DIR = f\"{DATA_DIR}/cellxgene\"\n",
    "\n",
    "NSFOREST_DIR = f\"{DATA_DIR}/nsforest-2024-06-27\"\n",
    "TOTAL_COUNTS = 5000  # TODO: Select a more sensible value\n",
    "\n",
    "NCBI_CELL_DIR = f\"{DATA_DIR}/ncbi-cell\"\n"
   ],
   "id": "8302423e-975e-4506-a5b9-42b1950c5c21"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we write the function, noting:\n",
    "\n",
    "-   Some datasets have multiple annotations per sample (ex.\n",
    "    \"broad<sub>celltype</sub>\" and \"granular<sub>celltype</sub>\").\n",
    "    NSForest can be run on multiple \\`cluster<sub>header</sub>\\`'s.\n",
    "    Combining the parent and child markers may improve classification\n",
    "    results.\n",
    "\n",
    "-   \\`adata.var<sub>names</sub>\\` must be unique. If there is a problem,\n",
    "    usually it can be solved by assigning \\`adata.var.index =\n",
    "    adata.var\\[\"ensembl<sub>id</sub>\"\\]\\`.\n",
    "\n",
    "-   Some datasets are too large and need to be downsampled to be run\n",
    "    through the pipeline. When downsampling, be sure to have all the\n",
    "    granular cluster annotations represented.\n",
    "\n",
    "-   Only run ns.pp.dendrogram() if there is no pre-defined dendrogram\n",
    "    order. This step can still be run with no effects, but the runtime\n",
    "    may increase."
   ],
   "id": "099415eb-fd7f-42a5-a6e2-fdf045eadfc8"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "results": "silent",
    "session": "shared",
    "tangle": "../py/NSForest.py"
   },
   "outputs": [],
   "source": [
    "def run_nsforest_on_file(h5ad_filename, cluster_header=\"cell_type\"):\n",
    "    \"\"\"Run NSForest using the specified dataset filename, and\n",
    "    cluster_header.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    h5ad_filename : str\n",
    "       The dataset filename\n",
    "    cluster_header : str\n",
    "       The cluster header\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    None\n",
    "    \"\"\"\n",
    "    # Assign results filename and directory\n",
    "    pp_h5ad_filename = f\"pp_{h5ad_filename}\"\n",
    "    results_dirname = h5ad_filename.split(\".\")[0]\n",
    "    results_dirpath = f\"{NSFOREST_DIR}/{results_dirname}\"\n",
    "\n",
    "    # Run NSForest if results do not exist\n",
    "    if not os.path.exists(results_dirpath):\n",
    "        os.makedirs(results_dirpath)\n",
    "\n",
    "        print(f\"Loading unprocessed AnnData file: {h5ad_filename}\")\n",
    "        h5ad_filepath = f\"{CELLXGENE_DIR}/{h5ad_filename}\"\n",
    "        up_adata = sc.read_h5ad(h5ad_filepath)\n",
    "\n",
    "        # TODO: Check validity of downsampling\n",
    "        print(\"Calculating QC metrics\")\n",
    "        up_metrics = sc.pp.calculate_qc_metrics(up_adata)\n",
    "        if up_metrics[1][\"total_counts\"].sum() > TOTAL_COUNTS:\n",
    "            print(\"Downsampling unprocessed AnnData file\")\n",
    "            ds_adata = sc.pp.downsample_counts(\n",
    "                up_adata, total_counts=TOTAL_COUNTS, copy=True\n",
    "            )\n",
    "        else:\n",
    "            ds_adata = up_adata  # No need to copy\n",
    "\n",
    "        print(\"Generating scanpy dendrogram\")\n",
    "        # Dendrogram order is stored in\n",
    "        # `pp_adata.uns[\"dendrogram_cluster\"][\"categories_ordered\"]`\n",
    "        pp_adata = up_adata.copy()\n",
    "        pp_adata.obs[cluster_header] = pp_adata.obs[cluster_header].astype(str)\n",
    "        pp_adata.obs[cluster_header] = pp_adata.obs[cluster_header].astype(\"category\")\n",
    "        pp_adata = ns.pp.dendrogram(\n",
    "            pp_adata,\n",
    "            cluster_header,\n",
    "            save=False,\n",
    "            output_folder=results_dirpath,\n",
    "            outputfilename_suffix=cluster_header,\n",
    "        )\n",
    "\n",
    "        print(\"Calculating cluster medians per gene\")\n",
    "        pp_adata = ns.pp.prep_medians(pp_adata, cluster_header)\n",
    "\n",
    "        print(\"Calculating binary scores per gene per cluster\")\n",
    "        pp_adata = ns.pp.prep_binary_scores(pp_adata, cluster_header)\n",
    "\n",
    "        pp_h5ad_filepath = f\"{results_dirpath}/{pp_h5ad_filename}\"\n",
    "        print(f\"Saving preprocessed AnnData file: {pp_h5ad_filepath}\")\n",
    "        pp_adata.write_h5ad(pp_h5ad_filepath)\n",
    "\n",
    "        print(f\"Running NSForest for preprocessed AnnData file: {pp_h5ad_filename}\")\n",
    "        results = nsforesting.NSForest(\n",
    "            pp_adata,\n",
    "            cluster_header,\n",
    "            output_folder=f\"{results_dirpath}/\",\n",
    "            outputfilename_prefix=cluster_header,\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        print(f\"Completed NSForest for preprocessed AnnData file: {pp_h5ad_filename}\")\n"
   ],
   "id": "d16d6cc6-f9eb-4af7-bd15-1c54bc2429d3"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now call the function for an example CELLxGENE dataset using the default\n",
    "`cluster_header` of `\"cell_type\"`:"
   ],
   "id": "cbf02a96-d711-4216-a8e4-56462af82f40"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "results": "output",
    "session": "shared"
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    h5ad_filename = \"6e00ccf7-0749-46ef-a999-dba785630d52.H5AD\"\n",
    "    run_nsforest_on_file(h5ad_filename, cluster_header=\"cell_type\")\n",
    "except Exception:\n",
    "    print_exc()\n"
   ],
   "id": "aa417b22-8dae-435a-b5cb-9a2ecf57d21f"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, in [Chapter-04-OntoGPT.ipynb](Chapter-04-OntoGPT.ipynb) we\n",
    "investigate the use of OntoGPT for publication processing. In\n",
    "[Chapter-05-ArangoDB.ipynb](Chapter-05-ArangoDB.ipynb) we'll use the\n",
    "results produced by NS-Forest to populate an ArangoDB database graph."
   ],
   "id": "fce10581-e5b6-427f-86de-61a36bf8b7fb"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
